src/conn/conn_prefetch.c

```diff
@@ -142,6 +141,7 @@ __wt_prefetch_thread_run(WT_SESSION_IMPL *session, WT_THREAD *thread)
         if (!F_ISSET(conn, WT_CONN_DATA_CORRUPTION) && pe->ref->page_del == NULL)
             WT_WITH_DHANDLE(session, pe->dhandle, ret = __wt_prefetch_page_in(session, pe));
 
+        F_CLR_ATOMIC_8(pe->ref, WT_REF_FLAG_PREFETCH);
```

Comment:
How this works with the following change?

Reply:
I agree that we are holding the flag for longer which means it gives a bigger time window for eviction to proceed with a leaf page that has the prefetch flag set.

Reply:
We are preventing evicting the internal page until we finish with the leaf page.

Reply:
OK, that is very subtle. Why we need to prevent evicting internal pages?

Reply:
We are holding a pointer to a `ref`, this is from the internal page memory. We need to prevent it from being freed until we finish with the `ref`.

Reply:
We remove the check in `__wt_page_can_evict` so I think we still can evict that page and the parent page. I don't think this prevents that completely. Even we do this, the ref we hold still may be outdated.

We evict the leaf with the flag and then parent.
We read the parent back.
We prefetch the leaf with the bad pointer.

Reply:
No. As you said, removing the check in `__wt_page_can_evict` can allow the leaf page to be evicted. This doesn't clear any flags that are set on the `ref`.  It just frees the `ref->page`. Due to the flags not being cleared, we cannot evict the internal page.

Reply:
```
It just frees the ref->page. Due to the flags not being cleared, we cannot evict the internal page.
```

We have evicted the leaf page and now if there is no leaf page on that internal page, what can stop us evicting the internal page?

Reply:
Discussed with @kommiharibabu, we have a check in __evict_child_check to ensure we don't evict the internal page. @etienneptl Can you add a comment to explain how the whole mechanism works?

Reply:
Also we should remove the locking in __evict_child_check

```
    /* Pre-fetch queue flags on a ref need to be checked while holding the pre-fetch lock. */
    __wt_spin_lock(session, &S2C(session)->prefetch_lock);
```

because we now use atomic operations on the flag.

Reply:
Removed in 270ed77615749341ed394fc6fc818a10d77a665c.

Reply:
And added a comment in 7eb97a6ec294e53afd2a18169c1843339f687cd3 @quchenhao

Reply:
Was it safe to remove the lock? Is there something else preventing a `WT_REF` that was already visited during `evict_child_check` having the prefetch flag set?

Reply:
That's a good point. It also raises another question. Even we don't remove the lock, we have a race like this:

* We scanned all the ref and there is no ref with the flag.
* We unlock the lock.
* Prefetch queue add one page of the internal page to the queue.
* We evict the internal page.
* We crash on the ref saved in the prefetch queue.

Reply:
Can we add an internal page if it's already in memory though?

Reply:
I think when we call `evict_child_check `, eviction has already taken exclusive access on the page so both the above scenarios should not happen.

Reply:
I am currently trying to reproduce a race issue but in theory, I think we could have an issue:

1. Thread 1 (T1) starts (`WT_INTL_FOREACH_BEGIN (session, parent->page, child)...)`
2. T1 goes through each child and `WT_REF_FLAG_PREFETCH` is not set
3. Another reader (T2) in parallel reads a leaf page that belongs to that same internal page. It adds other leaf pages to the prefetch queue.
4. T1 resumes and evicts the internal page
5. T2 processes the leaf page from the internal page that hs jsut been evicted by T1

What do you think @quchenhao?

Reply:
I think that race can happen even with the locks, if the leaf page is added after the loop encapsulated by `WT_INTL_FOREACH_BEGIN`.

Reply:
@etienneptl  The step 3 cannot happen because it needs to wait the eviction thread to give up exclusive access. After that, the page is evicted. It cannot do that in the middle of an eviction.

Reply:
I see, that's great. My reproducers are not hitting with the latest changes which include the lock removals.